qmt/geometry/geo_2d_data.py

```diff
@@ -132,10 +134,63 @@ def coord_list(self, part_name: str) -> List:
         elif isinstance(part, LineString):
             return list(np.array(part.coords.xy).T)[:]
 
+    def crop(
+        self,
+        x_min: float = None,
+        x_max: float = None,
+        y_min: float = None,
+        y_max: float = None,
+    ):
+        if x_min is not None and x_max is not None:
+            assert x_min < x_max, "x_max must be greater than x_min"
+        if y_min is not None and y_max is not None:
+            assert y_min < y_max, "y_max must be greater than y_min"
+        cropped_geo = Geo2DData(self.lunit)
+        x_min_old, x_max_old, y_min_old, y_max_old = self.compute_bb()
+        x_min = x_min_old if x_min is None else x_min
+        x_max = x_max_old if x_max is None else x_max
+        y_min = y_min_old if y_min is None else y_min
+        y_max = y_max_old if y_max is None else y_max
+        crop_poly = Polygon(
+            [[x_min, y_min], [x_min, y_max], [x_max, y_max], [x_max, y_min]]
+        )
+
+        ind_count = defaultdict(int)
+
+        for name, part in self.parts.items():
+            ind_name = name.split(":")[0] if ":" in name else name
+            cropped_part = crop_poly.intersection(part)
```

Comment:
We should probably have a test for the case where a part edge is co-linear with the bounding box edge.

Reply:
I've tested this and it works. The edge is included in the crop as a LineString. I think if you're using it for bdg all LineStrings are ignored, but from a geometry point of view it might be good to keep it there (for e.g. if you're setting boundary conditions for TF then it matters)

Reply:
I'll write a proper test for it as Andrey suggests though